components/fxa-client/src/state_machine/types.rs

```diff
@@ -0,0 +1,284 @@
+    Authenticating { oauth_url: String },
+    /// User is currently connected to FxA
+    Connected,
+    /// User was connected to FxA, but we observed issues with the auth tokens.
+    /// The user needs to reauthenticate before the account can be used.
+    AuthIssues,
+}
+
+/// Fxa event
+///
+/// These are the events that consumers send to [crate::FxaStateMachine::process_event]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum FxaEvent {
+    /// Initialize the state machine.  This must be the first event sent.
+    Initialize,
+    /// Begin an oauth flow
+    ///
+    /// If successful, the state machine will transition the [FxaState::Authenticating].  The next
+    /// step is to navigate the user to the `oauth_url` and let them sign and authorize the client.
+    BeginOAuthFlow { scopes: Vec<String> },
+    /// Begin an oauth flow using a URL from a pairing code
+    ///
+    /// If successful, the state machine will transition the [FxaState::Authenticating].  The next
+    /// step is to navigate the user to the `oauth_url` and let them sign and authorize the client.
+    BeginPairingFlow {
+        pairing_url: String,
+        scopes: Vec<String>,
+    },
+    /// Complete an OAuth flow.
+    ///
+    /// Send this event after the user has navigated through the OAuth flow and has reached the
+    /// redirect URI.  Extract `code` and `state` from the query parameters.  If successful the
+    /// state machine will transition to [FxaState::Connected].
+    CompleteOAuthFlow { code: String, state: String },
+    /// Cancel an OAuth flow.
+    ///
+    /// Use this to cancel an in-progress OAuth, returning to [FxaState::Disconnected] so the
+    /// process can begin again.
+    CancelOAuthFlow,
+    /// Check the authorization status for a connected account.
+    ///
+    /// Send this when issues are detected with the auth tokens for a connected account.  It will
+    /// double check for authentication issues with the account.  If it detects them, the state
+    /// machine will transition to [FxaState::AuthIssues].  From there you can start an OAuth flow
+    /// again to re-connect the user.
+    CheckAuthorizationStatus,
+    /// Disconnect the user
+    ///
+    /// Send this when the user is asking to be logged out.  The state machine will transition to
+    /// [FxaState::Disconnected].
+    Disconnect,
+}
+
+/// Internal [crate::FxaStateMachine] states
+///
+/// For each [FxaState] variant, there is a corresponding variant here.  These are called the
+/// public states since they are visible to consumer applications.
+///
+/// There are also variants for internal states.  This states indicate that the state machine is
+/// making a [crate::internal::FirefoxAccount] call. Internal states are temporary.
+/// [crate::FxaStateMachine] will transition to a new state once the call is complete and
+/// before `process_event` completes.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum State {
```

Comment:
I initially had a less flattened version of this:

```
pub enum State {
  Public(FxaState),
  Internal(InternalState),
}

pub enum InternalState {
  GetAuthState,
  BeginOAuthFlow { .. },
  ...
}
```

From the view of the types I liked that better, but it made the logic hard to follow because everything was so verbose.